crates/router/src/core/utils.rs

```diff
@@ -1073,3 +1079,57 @@ pub fn get_flow_name<F>() -> RouterResult<String> {
         .attach_printable("Flow stringify failed")?
         .to_string())
 }
+
+pub async fn persist_individual_surcharge_details_in_redis(
+    state: &AppState,
+    merchant_account: &domain::MerchantAccount,
+    surcharge_metadata: &api_models::payment_methods::SurchargeMetadata,
+) -> RouterResult<()> {
+    let redis_conn = state
+        .store
+        .get_redis_conn()
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("Failed to get redis connection")?;
+
+    let redis_store_futures_iter = surcharge_metadata
+        .get_individual_surcharge_key_value_pairs()
+        .into_iter()
+        .map(|(redis_key, surcharge_details)| {
+            redis_conn.serialize_and_set_key_with_expiry(
+                redis_key,
+                surcharge_details,
+                merchant_account
+                    .intent_fulfillment_time
+                    .unwrap_or(DEFAULT_FULFILLMENT_TIME),
```

Comment:
use `set_hash_fields` redis function here. The key would be `surcharge_metadata_<attempt_id>` and the individual fields can be hashmap keys.